kv/kvserver: TestReadLoadMetricAccounting failed

kv/kvserver.TestReadLoadMetricAccounting [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12187970?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12187970?buildTab=artifacts#/) on release-23.2 @ [87a94f6ac59d4e01c822bb2128db858176600dd6](https://github.com/cockroachdb/cockroach/commits/87a94f6ac59d4e01c822bb2128db858176600dd6):


```
=== RUN   TestReadLoadMetricAccounting
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReadLoadMetricAccounting3976845280
    test_log_scope.go:81: use -show-logs to present logs inline
    replica_rankings_test.go:210: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rankings_test.go:210
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rankings_test.go:433
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:75
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:81
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:60
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:42
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rankings_test.go:398
        	Error:      	Max difference between 0 and 79 allowed is 60, but difference was -79
        	Test:       	TestReadLoadMetricAccounting
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReadLoadMetricAccounting3976845280
--- FAIL: TestReadLoadMetricAccounting (8.38s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadLoadMetricAccounting.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-32393

--- a/pkg/kv/kvserver/replica_rankings_test.go
+++ b/pkg/kv/kvserver/replica_rankings_test.go
@@ -322,14 +322,11 @@ func TestReadLoadMetricAccounting(t *testing.T) {
 	const epsilonAllowed = 4
 
 	scratchKey := tc.ScratchRange(t)
-	nextKey := scratchKey.Next()
-
-	scratchKeys := make([]roachpb.Key, 300)
 	sstKeys := make(storageutils.KVs, 300)
 	for i := range sstKeys {
-		scratchKeys[i] = nextKey
-		sstKeys[i] = storageutils.PointKV(nextKey.String(), 1, "value")
-		nextKey = nextKey.Next()
+		// Format each key as /Table/Max/000000001, /Table/Max/000000002, etc.
+		sstKeys[i] = storageutils.PointKV(
+			string(scratchKey)+fmt.Sprintf("%09d", i), 1, "value")
 	}
 	sst, start, end := storageutils.MakeSST(t, ts.ClusterSettings(), sstKeys)
 	sstReq := &kvpb.AddSSTableRequest{
@@ -359,6 +356,10 @@ func TestReadLoadMetricAccounting(t *testing.T) {
 	scanReadBA := &kvpb.BatchRequest{}
 	scanReadBA.Add(scan)
 
+	// NB: Each KV pair is an identical size, 38 bytes. Expect that as the
+	// number of keys being read increases, the read bytes scales by keys * 38.
+	const entrySize = 38
+
 	testCases := []struct {
 		ba           *kvpb.BatchRequest
 		expectedRQPS float64
@@ -368,73 +369,57 @@ func TestReadLoadMetricAccounting(t *testing.T) {
 		expectedRBPS float64
 	}{
 		{getReadBA, 1, 0, 1, 0, 10},
-		{genVariableRead(ctx, start, sstKeys[1].(storage.MVCCKeyValue).Key.Key), 1, 0, 1, 0, 38},
-		{genVariableRead(ctx, start, sstKeys[4].(storage.MVCCKeyValue).Key.Key), 1, 0, 4, 0, 176},
-		{genVariableRead(ctx, start, sstKeys[64].(storage.MVCCKeyValue).Key.Key), 1, 0, 64, 0, 10496},
+		{genVariableRead(ctx, start, sstKeys[1].(storage.MVCCKeyValue).Key.Key), 1, 0, 1, 0, entrySize},
+		{genVariableRead(ctx, start, sstKeys[4].(storage.MVCCKeyValue).Key.Key), 1, 0, 4, 0, 4 * entrySize},
+		{genVariableRead(ctx, start, sstKeys[64].(storage.MVCCKeyValue).Key.Key), 1, 0, 64, 0, 64 * entrySize},
 	}
 
 	store, err := ts.GetStores().(*Stores).GetStore(ts.GetFirstStoreID())
 	require.NoError(t, err)
 
+	// Ensure that the all the keys in the AddSST sent were for the same range.
 	repl := store.LookupReplica(roachpb.RKey(start))
 	require.NotNil(t, repl)
 
 	replEnd := store.LookupReplica(roachpb.RKey(end))
 	require.NotNil(t, repl)
 
-	require.EqualValues(t, repl, replEnd)
+	require.EqualValues(t, repl.Desc().RangeID, replEnd.Desc().RangeID)
 
 	// Disable the consistency checker, to avoid interleaving requests
 	// artificially inflating measurement due to consistency checking.
 	sqlDB.Exec(t, `SET CLUSTER SETTING server.consistency_check.interval = '0'`)
-	sqlDB.Exec(t, `SET CLUSTER SETTING kv.range_split.by_load.enabled = false`)
 
 	for _, testCase := range testCases {
-		// This test can flake, where an errant request - not sent here
-		// (commonly intent resolution) will artifically inflate the collected
-		// metrics. This results in unexpected read/write statistics and a
-		// flakey test every few hundred runs. Here we assert that the run
-		// should succeed soon, if it fails on the first.
-		testutils.SucceedsSoon(t, func() error {
-			// Reset the request counts to 0 before sending to clear previous requests.
-			// Reset the request counts to 0 before sending to clear previous requests.
-			repl.loadStats.Reset()
+		// Reset the request counts to 0 before sending to clear previous requests.
+		repl.loadStats.Reset()
 
-			requestsBefore := repl.loadStats.TestingGetSum(load.Requests)
-			writesBefore := repl.loadStats.TestingGetSum(load.WriteKeys)
-			readsBefore := repl.loadStats.TestingGetSum(load.ReadKeys)
-			readBytesBefore := repl.loadStats.TestingGetSum(load.ReadBytes)
-			writeBytesBefore := repl.loadStats.TestingGetSum(load.WriteBytes)
+		requestsBefore := repl.loadStats.TestingGetSum(load.Requests)
+		writesBefore := repl.loadStats.TestingGetSum(load.WriteKeys)
+		readsBefore := repl.loadStats.TestingGetSum(load.ReadKeys)
+		readBytesBefore := repl.loadStats.TestingGetSum(load.ReadBytes)
+		writeBytesBefore := repl.loadStats.TestingGetSum(load.WriteBytes)
 
-			_, pErr = db.NonTransactionalSender().Send(ctx, testCase.ba)
-			require.Nil(t, pErr)
+		_, pErr = db.NonTransactionalSender().Send(ctx, testCase.ba)
+		require.Nil(t, pErr)
 
-			require.Equal(t, 0.0, requestsBefore)
-			require.Equal(t, 0.0, writesBefore)
-			require.Equal(t, 0.0, readsBefore)
-			require.Equal(t, 0.0, writeBytesBefore)
-			require.Equal(t, 0.0, readBytesBefore)
+		require.Equal(t, 0.0, requestsBefore)
+		require.Equal(t, 0.0, writesBefore)
+		require.Equal(t, 0.0, readsBefore)
+		require.Equal(t, 0.0, writeBytesBefore)
+		require.Equal(t, 0.0, readBytesBefore)
 
-			requestsAfter := repl.loadStats.TestingGetSum(load.Requests)
-			writesAfter := repl.loadStats.TestingGetSum(load.WriteKeys)
-			readsAfter := repl.loadStats.TestingGetSum(load.ReadKeys)
-			readBytesAfter := repl.loadStats.TestingGetSum(load.ReadBytes)
-			writeBytesAfter := repl.loadStats.TestingGetSum(load.WriteBytes)
+		requestsAfter := repl.loadStats.TestingGetSum(load.Requests)
+		writesAfter := repl.loadStats.TestingGetSum(load.WriteKeys)
+		readsAfter := repl.loadStats.TestingGetSum(load.ReadKeys)
+		readBytesAfter := repl.loadStats.TestingGetSum(load.ReadBytes)
+		writeBytesAfter := repl.loadStats.TestingGetSum(load.WriteBytes)
 
-			assertGreaterThanInDelta(t, testCase.expectedRQPS, requestsAfter, epsilonAllowed)
-			assertGreaterThanInDelta(t, testCase.expectedWPS, writesAfter, epsilonAllowed)
-			assertGreaterThanInDelta(t, testCase.expectedRPS, readsAfter, epsilonAllowed)
-			// Increase the allowance for write bytes, as although this should
-			// be zero - there exists some cases where a write will interleave
-			// with our testing. This causes the test to flake if it occurs repeatedly.
-			// TODO(kvoli): This can be determinsitic, with an general
-			// interface that records the type of request or otherwise filters
-			// out unwanted types.
-			assertGreaterThanInDelta(t, testCase.expectedWBPS, writeBytesAfter, epsilonAllowed*15)
-			assertGreaterThanInDelta(t, testCase.expectedRBPS, readBytesAfter, epsilonAllowed)
-
-			return nil
-		})
+		assertGreaterThanInDelta(t, testCase.expectedRQPS, requestsAfter, epsilonAllowed)
+		assertGreaterThanInDelta(t, testCase.expectedWPS, writesAfter, epsilonAllowed)
+		assertGreaterThanInDelta(t, testCase.expectedRPS, readsAfter, epsilonAllowed)
+		assertGreaterThanInDelta(t, testCase.expectedWBPS, writeBytesAfter, epsilonAllowed)
+		assertGreaterThanInDelta(t, testCase.expectedRBPS, readBytesAfter, epsilonAllowed)
 	}
 }
 

